test/test-users.js

```diff
@@ -0,0 +1,402 @@
+            res.charset.should.equal('utf-8');
+            res.body.should.be.an('object');
+            res.body.should.have.property('message');
+            res.body.name.should.equal('ValidationError');
+            res.body.message.should.equal('User validation failed');
+            res.body.errors.password.message.should.equal('Path `password` is required.');
+          })
+          .then(() => {
+            spy.should.not.have.been.called();
+          });
+      });
+      it('should reject duplicate entry for johnDoe', () => {
+        let spy = chai.spy();
+        return chai.request(app).post('/users').send(johnDoe)
+          .then(spy)
+          .catch(err => {
+            // If the request fails, make sure it contains the error
+            let res = err.response;
+            res.should.have.status(409);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+            res.body.should.be.an('object');
+
+            res.body.should.have.property('name');
+            res.body.name.should.equal('MongoError');
+            res.body.code.should.equal(11000); // E11000 duplicate key error
+          })
+          .then(() => {
+            // Check that the request didn't succeed
+            spy.should.not.have.been.called();
+          });
+      });
+    });
+
+    describe('/users/me', function() {
+      describe('GET', function() {
+        it('should 401 on un-authenticated user', function() {
+          let spy = chai.spy();
+          // this call does not attempt to authenticate
+          return chai.request(app).get('/users/me')
+            .then(spy)
+            .catch(function(err) {
+              err.response.should.have.status(401);
+            })
+            .then(function() {
+              spy.should.not.have.been.called();
+            });
+        });
+        it('should 401 on failed username authentication', function() {
+          let spy = chai.spy();
+          // this call attempts a BAD username and good passowrd
+          return chai.request(app).get('/users/me').auth('unknown', johnDoe.password)
+            .then(spy)
+            .catch(function(err) {
+              err.response.should.have.status(401);
+            })
+            .then(function() {
+              spy.should.not.have.been.called();
+            });
+        });
+        it('should 401 on failed password authentication', function() {
+          let spy = chai.spy();
+          // this call attempts a good username and BAD passowrd
+          return chai.request(app).get('/users/me').auth(johnDoe.username, 'bad-password')
+            .then(spy)
+            .catch(function(err) {
+              err.response.should.have.status(401);
+            })
+            .then(function() {
+              spy.should.not.have.been.called();
+            });
+        });
+        it('successful authentication should return a single user', () => {
```

Comment:
prefer to have this come first, just so that conceptually, we're always testing normal case first, then a handful of edge cases second.